Feat (mfusg): FloPy MODFLOW-USG support (with .bas file example) #1476

swfwmd opened the pull request:
* Feat(mfusg): add mfusg transport version 2.4.0 support

fixes and updates included with feature:
New Features
* MfUsgBct : Block Centered Transport (BCT)
* MfUsgPcb : Prescribed Concentration Boundary (PCB)
* MfUsgDdf: Density Driven Flow (DDF)
* MfUsgDpf: Dual Porosity Flow (DPF)
* MfUsgDpt: Dual Porosity Transport (DPT)
* MfUsgMdt: Matrix Diffusion Transport (MDT)
* MfUsgRch: Recharge (RCH)
* MfUsgLak: Lake (LAK)

Modified Packages
* MfUsgBcf : Block Centered Flow (BCF)
* MfUsgCln :  Connected Linear Network (CLN)
* MfUsgSms :  Sparse Matrix Solver (SMS)
* MfUsgWel :  Well (WEL)
* ModflowFhb : Flow and Head Boundary (FHB)

See Panday, S., 2024; USG-Transport Version 2.4.0: Transport and Other Enhancements to MODFLOW-USG, GSI Environmental, July 2024 http://www.gsi-net.com/en/software/free-software/USG-Transport.html

hzhang4 committed:
Feat (mfusg): FloPy MODFLOW-USG support (with .bas file example) #1476

* Feat(mfusg): add mfusg transport version 2.4.0 support

fixes and updates included with feature:
New Features
* MfUsgBct : Block Centered Transport (BCT)
* MfUsgPcb : Prescribed Concentration Boundary (PCB)
* MfUsgDdf: Density Driven Flow (DDF)
* MfUsgDpf: Dual Porosity Flow (DPF)
* MfUsgDpt: Dual Porosity Transport (DPT)
* MfUsgMdt: Matrix Diffusion Transport (MDT)
* MfUsgRch: Recharge (RCH)
* MfUsgLak: Lake (LAK)

Modified Packages
* MfUsgBcf : Block Centered Flow (BCF)
* MfUsgCln :  Connected Linear Network (CLN)
* MfUsgSms :  Sparse Matrix Solver (SMS)
* MfUsgWel :  Well (WEL)
* ModflowFhb : Flow and Head Boundary (FHB)

See Panday, S., 2024; USG-Transport Version 2.4.0: Transport and Other Enhancements to MODFLOW-USG, GSI Environmental, July 2024 http://www.gsi-net.com/en/software/free-software/USG-Transport.html

langevin-usgs commented:
Thanks for this, @swfwmd.  Have you thought about some of the other issues raised in #1476?  It's still unclear how to test these changes.  Maybe @cnicol-gwlogic or @martinvonk have some thoughts?

swfwmd commented:
I am working on the test cases and will commit later

cnicol-gwlogic commented:
My understanding was a key barrier to flopy tests for usg-transport was this: a git repo with tagged releases and ci testing for usg-transport was needed, so the flopy ci can pull in that executable as needed as usg-transport is developed. 

I had a very quick stab at this some time ago based on mf6, but again haven't had the time to finish it. It needs some work.

martinvonk commented:
> My understanding was a key barrier to flopy tests for usg-transport was this: a git repo with tagged releases and ci testing for usg-transport was needed, so the flopy ci can pull in that executable as needed as usg-transport is developed.

This was also my understanding. And I did not have the time (or fortran experience :)) either to set this up. So if we'd want this, we would have to:
1. Set up and improve the [USG-transport](https://github.com/cnicol-gwlogic/usg-transport) repo with tests, releases and all the fancy CI stuff
2. Merge the two FloPy forks that now exist for MODFLOW USG-Transport. For example, my [fork](https://github.com/martinvonk/flopy/tree/usg) has the Richards equation and PFAS transport options. Haven't looked closely at @swfwmd fork yet for it's functionalities. 
3. Make tests and example notebooks for the FloPy usage of MODFLOW USG-Transport

hzhang4 commented:
I will contact Sorab to see if it is possible to set up a git repo for usg-transport.

jdhughes-usgs commented:
The gsi version of mfusg is available from https://github.com/MODFLOW-USGS/executables/releases. This could be used to test usg-transport flopy functionality.

hzhang4 committed:
Feat (mfusg): FloPy MODFLOW-USG support (with .bas file example) modflowpy#1476
* Feat(mfusg): add mfusg transport version 2.4.0 support

fixes and updates included with feature:
New Features
* MfUsgEvt: EVT added ETFACTOR support for transport

Modified Packages
* MfUsg :
* MfUsgBct :
* MfUsgCln :
* MfUsgWel :
* MfUsgDpf :
* MfUsgRch :

Examples:
Added usg_tranport folder under example\data
with following examples:
Ex1-1D
Ex2_Radial_2D
Ex3_CLN_Conduit
Ex4_Dual_Domain
Ex5_Henry
Ex6_Stallman
Ex7_Multispecies
Ex8_Lake
Ex9_PFAS

See Panday, S., 2024; USG-Transport Version 2.4.0: Transport and Other Enhancements to MODFLOW-USG, GSI Environmental, July 2024 http://www.gsi-net.com/en/software/free-software/USG-Transport.html

swfwmd committed:
Fixed issues with the test_usg_transport.py

hzhang4 committed:
Fixed issues with the test_usg_transport.py and test_usg.py for pytest

hzhang4 committed:
Feat (mfusg): FloPy MODFLOW-USG support (with .bas file example) #1476
Hua Zhang @ swfwmd
Feat(mfusg): add mfusg transport version 2.3.0 support
fixes and updates included with feature: New Features

MfUsgBct : Block Centered Transport (BCT)
MfUsgPcb : Prescribed Concentration Boundary (PCB)
MfUsgDdf: Density Driven Flow (DDF)
MfUsgDpf: Dual Porosity Flow (DPF)
MfUsgDpt: Dual Porosity Transport (DPT)
MfUsgMdt: Matrix Diffusion Transport (MDT)
MfUsgRch: Recharge (RCH)
MfUsgLak: Lake (LAK)
MfUsgEvt: Evt
Modified Packages

MfUsgBcf : Block Centered Flow (BCF)
MfUsgCln : Connected Linear Network (CLN)
MfUsgSms : Sparse Matrix Solver (SMS)
MfUsgWel : Well (WEL)
ModflowFhb : Flow and Head Boundary (FHB)
ModflowChd : CHD
ModflowBas : BAS

Auto test:
Added test_usg_transport.py
Added example/data/mfusg_transport with 9 test examples
Added example notebooks
.docs/Notebooks/mfusg_transport_Ex9_PFAS.py .docs/Notebooks/mfusg_transport_Ex1_1D.py .docs/Notebooks/mfusg_transport_Ex2_Radial.py .docs/Notebooks/mfusg_transport_Ex3_Conduit.py .docs/Notebooks/mfusg_transport_Ex4_DualDomain.py .docs/Notebooks/mfusg_transport_Ex5_Henry.py .docs/Notebooks/mfusg_transport_Ex6_Stallman_Heat_Transport.py .docs/Notebooks/mfusg_transport_Ex7a_Discrete_Fracture.py .docs/Notebooks/mfusg_transport_Ex7b_SandBox.py .docs/Notebooks/mfusg_transport_Ex7c_ChainDecay.py .docs/Notebooks/mfusg_transport_Ex8_Lake.py

Modified test_usg.py

See Panday, S., 2024; USG-Transport Version 2.4.0: Transport and Other Enhancements to MODFLOW-USG, GSI Environmental, July 2024 http://www.gsi-net.com/en/software/free-software/USG-Transport.html

hzhang4 committed:
Merge branch 'develop' of https://github.com/modflowpy/flopy into modflowpy-develop

hzhang4 committed:
Merge branch 'modflowpy-develop' into mfusg

swfwmd commented:
Added tests and notebooks:

Added test_usg_transport.py
Added example/data/mfusg_transport
with 9 test examples
Ex1_1D
Ex2_Radial
Ex3_Conduit
Ex4_DualDomain
Ex5_Henry
Ex6_Stallman
Ex7_Matrix_Diffusion
Ex8_Lake
Ex9_PFAS

Added example notebooks
mfusg_transport_Ex1_1D.py
mfusg_transport_Ex2_Radial.py
mfusg_transport_Ex3_Conduit.py
mfusg_transport_Ex4_DualDomain.py
mfusg_transport_Ex5_Henry.py
mfusg_transport_Ex6_Stallman_Heat_Transport.py
mfusg_transport_Ex7a_Discrete_Fracture.py
mfusg_transport_Ex7b_SandBox.py
mfusg_transport_Ex7c_ChainDecay.py
mfusg_transport_Ex8_Lake.py
mfusg_transport_Ex9_PFAS.py

Modified test_usg.py

swfwmd committed:
Update __init__.py

to include "MfUsgWel",

hzhang4 committed:
Clean up example\data\mfusg_transport folder. The examples were tested using mfusg transport 2.5.0

hzhang4 committed:
modified autotest and notebook files for mfusg transport for ruff and pytest workflow

swfwmd committed:
Merge branch 'develop' into mfusg

swfwmd committed:
Update test_usg.py

swfwmd committed:
Update test_usg.py

hzhang4 committed:
ruff check fix

hzhang4 committed:
ruff format

hzhang4 committed:
pytest fixed

hzhang4 committed:
Update mfusg_transport notebooks

wpbonelli commented:
is this PR still relevant, or did #2575 supersede it?
